core/src/main/java/org/apache/iceberg/io/ResolvingFileIO.java

```diff
@@ -83,6 +86,29 @@ public void deleteFile(String location) {
     io(location).deleteFile(location);
   }
 
+  @Override
+  public void deleteFiles(Iterable<String> pathsToDelete) throws BulkDeletionFailureException {
+    Map<FileIO, List<String>> pathByFileIO =
+        StreamSupport.stream(pathsToDelete.spliterator(), false)
```

Comment:
could be expensive, but I would hope even at 20 million 100 character strings that only sets us back about 2GB? I think for users in that situation they probably should just use a much larger driver?

Reply:
We could add batches here also, e.g. using `Iterators.partition()`. We could set the batch size to something large.

Reply:
I think we can do batch but I realized we ready have a batch size as properties defined on concrete fileIO, here's one example for S3: https://github.com/apache/iceberg/blob/29eebdd89efdcb9fb32e53325dfd2b29b9a3377b/aws/src/main/java/org/apache/iceberg/aws/s3/S3FileIOProperties.java#L209-L217. Shall we introduce a new set of configuration to batch the request or provide some fixed batch size and allow for customization later on?

Reply:
Yeah you'd be batching twice, once at this layer, once at the delegate. But you could set the batch size here to be something very large, just as a sanity limit to prevent OOMing. I feel even a large fixed size is better than unlimited, say 100k. Both GCP and AWS have much smaller batch limits so batching will still be done there.

Reply:
thank you @bryanck , added default 100k batch size per your suggestion